packages/auth/lib/Fields.js

```diff
@@ -0,0 +1,70 @@
+const moment = require('moment')
+
+const { newAuthError } = require('./AuthError')
+const Roles = require('./Roles')
+
+const MissingFieldsError = newAuthError(
+  'missing-fields',
+  'api/fields/missing'
+)
+
+const getMissingFields = async ({ user, email, pgdb }) => {
+  const missingFields = []
+  const unassignedGrants = await pgdb.public.accessGrants.find({
+    email,
+    recipientUserId: null,
+    'beginAt <=': moment(),
+    'endAt >': moment(),
+    invalidatedAt: null
+  })
+
+  const isMember = !!user && Roles.userHasRole(user, 'member')
+  const hasNames = !!user && (
+    user.firstName &&
+    user.firstName.trim().length > 1 &&
+    user.lastName &&
+    user.lastName.trim().length > 1
+  )
+  const hasGrants = unassignedGrants.length > 0
+
+  if (
+    (user && isMember && !hasNames) ||
+    (!user && hasGrants)
+  ) {
+    missingFields.push('firstName')
+    missingFields.push('lastName')
+  }
+
+  return missingFields
+}
+
+const ensureRequiredFields = async ({ user, email, providedFields, pgdb }) => {
+  const missingFields = []
+  const fields = {}
+
+  const expectedFields = await getMissingFields({ user, email, pgdb })
+
+  expectedFields
+    .forEach(expectedField => {
+      if (!providedFields[expectedField]) {
+        missingFields.push(expectedField)
+        return false
+      }
+
+      fields[expectedField] = providedFields[expectedField]
```

Comment:
IMO this is a bad approach: if one sends fields but they are not missing on the user they are ignored without a warning.

You don't have to worry about users sending additional fields through the mutation resolver:
```gql
mutation {
  authorizeSession(
    email: "test",
    tokens: [{type: EMAIL_TOKEN, payload: "lala"}],
    fields: {
      firstName: "Troll",
      lastName: "Leet",
      birthday: "dead"
    }
  )
}
```
will result in
```json
{
  "errors": [
    {
      "message": "Field \"birthday\" is not defined by type RequiredUserFields.",
      "locations": [
        {
          "line": 8,
          "column": 7
        }
      ]
    }
  ]
}
```

Reply:
I understand.

Did not worry about additional fields, fully aware GraphQL schema would return an error if undefined.

Thinking behind this flow was this:

You can only update fields what `unauthorizedSession` lists in `requiredFields`. This would prevent the ability to update first and last name "unwarranted" during `authorizeSession`.

I still believe that is a healthy, safe way. We could throw an error if there are not required fields provided. This or process whatever is provided.

Not feeling strong about it.

Reply:
Ok but then call the argument `requiredFields` in the `authorizeSession` mutation, that way there is a clear hint that it will only do when previously required.